bluemira/magnetostatics/circuits.py

```diff
@@ -33,10 +35,16 @@ class PlanarCircuit(SourceGroup):
     Base class for a planar current loop
     """
 
-    shape: np.ndarray | Coordinates
+    shape: npt.NDArray[np.float64] | Coordinates
     current: float
 
-    def _generate_sources(self, shape, current, source_class, xs_args):
+    def _generate_sources(
+        self,
+        shape: npt.NDArray[np.float64] | Coordinates,
+        current: float,
+        source_class: Any,
+        xs_args: npt.NDArray[np.float64],
+    ) -> list[Any]:
```

Comment:
Any is bad don't use Any where possible, is there either a abstract baseclass or a protocol we can create?

Reply:
Yeah I tried a couple of things but got a variety of errors. Do I have to go all the way to the baseclass for that?

Reply:
So it need to include every possible source class (eg the base class or a common ancestor) or a protocol which implements stubs of the methods a source will have (see your next papercut for an example)

Reply:
So when I use a common ancestor for PolyhedralPrismCurrentSource and TrapezoidalPrismCurrentSource which is CrossSectionCurrentSource it throws the error that it is not callable and also that "Type[TrapezoidalPrismCurrentSource]" doesn't match "CrossSectionCurrentSource"

Reply:
You need as the type `type[CrossSectionCurrentSource]` because you are passing in an unitialised class

Reply:
Ah I was using type() before. That makes a difference, although it seems to have a weird issue...  error: "TrapezoidalPrismCurrentSource" gets multiple values for keyword argument "alpha",

Reply:
resolved offline